packages/dapper/src/queryString.ts

```diff
@@ -0,0 +1,125 @@
+import { isNumberArray, isStringArray } from "./typeguards";
+
+type QueryStringable = string | number | boolean | string[] | number[];
+
+interface QsItem<T extends QueryStringable> {
+  key: string;
+  value?: T;
+  impotent?: T;
+}
+
+interface DefinedQsItem<T extends QueryStringable> extends QsItem<T> {
+  value: T;
+}
+
+export type QsArray = (
+  | QsItem<string>
+  | QsItem<number>
+  | QsItem<boolean>
+  | QsItem<string[]>
+  | QsItem<number[]>
+)[];
+
+/**
+ * Create a cache-friendly query string from received objects.
+ *
+ * To increase the likelyhood of cache hits
+ * * Empty values are filtered out.
+ * * Impotent values are filtered out. Impotent values are ones that
+ *   would have no effect on the resulting response if submitted to
+ *   backend, e.g. default values.
+ * * Parameter keys and values are sorted alphabetically.
+ *
+ * Accepts an array of objects containing the following fields, where
+ * `T = string | number | boolean | string[] | number[]`:
+ * * `key` (string): query string parameter name
+ * * `value?` (T): query string parameter value
+ * * `impotent?` (T): impotent value for this query string parameter
+ */
+export const serializeQueryString = (allParams: QsArray): string => {
+  const keys = allParams.map((p) => encodeURIComponent(p.key));
+
+  if (keys.length != new Set(keys).size) {
+    throw new Error("Duplicate keys are not allowed. Use value array instead.");
+  }
+
+  return (
+    _filterEmptyAndImpotent(allParams)
+      // Force lower case to improve the changes of cache hits.
```

Comment:
Typo (`changes` -> `chances`), but also are we sure this is a valid way of handling query strings? More specifically, do we have guarantees our query parameters are in fact always case insensitive? Same for values, seems like a risky assumption to make.

Reply:
IDK, what kinda use cases do we have for supporting case insensitive values? I mean, sure we can not lower case the values if you feel like keeping up the option of possibly having case insensitive values outweights the possibility of cache misses. Doign the same for keys poses the additional problem of detecting duplicates/arrays.

Fixed the typos.

Reply:
> what kinda use cases do we have for supporting case insensitive values?

I assume you mean **case-sensitive**, as the current proposed implementation would make them case-insensitive. For an example where case-insensitive values don't work, string based IDs that use lower & upper chars are fairly common (e.g. base64-encoded IDs which aren't too uncommon), which would break if we implicitly cast everything to lowercase.

In terms of caching, if our server is treating query params as case-sensitive, incorrect casings would lead to error pages and thus wouldn't be a problem for the cache.

So I believe we should not implicitly transform the values in any way since that might be significant to the functionality. Ordering is not significant (at least in any of the current APIs we have, and I'd imagine it's not going to be in future APIs either).

Reply:
> I assume you mean case-sensitive

Yep yep yep.

Casing is no longer adjusted for neither keys nor the values. Also removed the related test case.